File: backend/src/main/java/com/tmk/vtcmanager/application/usecases/contravention/RestaurerContraventionUseCase.java
package com.tmk.vtcmanager.application.usecases.contravention;

import com.tmk.vtcmanager.application.domain.contravention.Contravention;
import com.tmk.vtcmanager.application.exception.ResourceNotFoundException;
import com.tmk.vtcmanager.application.ports.persistence.ContraventionRepository;
import com.tmk.vtcmanager.application.services.VerrouArreteService;
import lombok.RequiredArgsConstructor;
import org.springframework.transaction.annotation.Transactional;

/**
 * Remet en circulation une contravention annulée à tort.
 *
 * <p>Elle retrouve le statut que dicte ce que le chauffeur a versé — en attente
 * si rien n'a été payé — et son marquage d'annulation s'efface : la créance
 * redevient exigible.
 *
 * <p>Refusé une fois la période de l'infraction clôturée : les états du mois ont
 * été arrêtés sans cette créance, l'y remettre après coup les ferait mentir.
 */
@RequiredArgsConstructor
public class RestaurerContraventionUseCase {

    private final ContraventionRepository contraventionRepository;
    private final VerrouArreteService verrouArreteService;

    @Transactional
    public Contravention execute(Long id) {
        Contravention contravention = contraventionRepository.findById(id)
                .orElseThrow(() -> ResourceNotFoundException.of("Contravention", id));

        if (!contravention.estAnnulee()) {
            throw new IllegalStateException(
                    "Cette contravention n'est pas annulée : rien à restaurer.");
        }
        verrouArreteService.verifier(contravention.getDateInfraction());

        contravention.restaurer();
        return contraventionRepository.save(contravention);
    }
}

File: mobile/lib/features/contravention/domain/usecases/restaurer_contravention_usecase.dart
import 'package:fpdart/fpdart.dart';

import '../../../../core/error/failure.dart';
import '../entities/contravention.dart';
import '../repositories/contravention_repository.dart';

/// Remet en circulation une contravention annulée à tort.
///
/// Elle retrouve le statut que dicte ce que le chauffeur a versé — en attente
/// si rien n'a été payé — et redevient exigible. Le serveur la refuse une fois
/// la période comptable clôturée : le mois a été arrêté sans cette créance.
class RestaurerContraventionUseCase {
  final ContraventionRepository _repository;
  const RestaurerContraventionUseCase(this._repository);

  Future<Either<Failure, Contravention>> call(int id) =>
      _repository.restaurerContravention(id);
}
